Copy plain text variant of blocks

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR strips the HTML from copied block markup for the text/plain version of the clipboard.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Pasting into plain text editors should only paste the text we see on screen not the whole underlying 
markup.

This is an improvement in the experience of writing in a different app rather than in the 
WordPress editor itself, and then bringing over the content. Since one could move text content back 
and forth a few times, pasting block markup is annoying if not even prone to introducing problems.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

By fixing what seems to be a problem anyway: stripping HTML out of serialized data in the copy 
handler before placing it in the `text/plain` version.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Switch to this PR
2. Create a post
3. Copy the post content
4. Paste in a plain text editor such as Notepat or Textedit (for Textedit make sure to use Format > 
Make plain text before pasting)
5. Notice there is no block markup pasted

## Screenshots or screencast <!-- if applicable -->

### Before

https://user-images.githubusercontent.com/107534/170446904-30d7883c-8449-40cc-b86e-89711a7bbd57.mp4

### After

https://user-images.githubusercontent.com/107534/170446592-01c3df9b-7e3f-4a80-a844-b99b135add2c.mp4

draganescu committed:
copy plain text into the clipboard by stripping HTML from serialized block data

mtias commented:
cc @mcsf who has thought about these mechanics before

mcsf commented on packages/block-editor/src/components/copy-handler/index.js:
```diff
@@ -158,7 +159,10 @@ export function useClipboardHandler() {
-					event.clipboardData.setData( 'text/plain', serialized );
+					event.clipboardData.setData(
+						'text/plain',
+						stripHTML( serialized )
```
If we go forward with this, we'll want to clean up line breaks:

```suggestion
						stripHTML( serialized ).replace( /\n\n+/g, '\n\n' )
```

I'm trying to decide whether this belongs inside `stripHTML`; it's a question of whether it can break any expectations within Gutenberg Core, and — although it's an explicitly `__unstable` API — any expectations with third parties.

mcsf reviewed (commented):
The value in this change is clear, but we need to identify the use cases for stripped and unstripped content copying, so as to minimise the disruption for users who have come to expect this behaviour. :)

Btw, I quickly wrote and uploaded a "clipboard debugger" to dive into this topic: http://mcsf-clipboard-data.surge.sh/ . This works as expected in Firefox, but I see strange data in Safari, which I suspect to be due to some policy of per-domain clipboard isolation or similar.

So, coming back to identifying use cases, we have:

1. Select a range of blocks and copy with <kbd>Meta+C</kbd> — handled by [CopyHandler](https://github.com/WordPress/gutenberg/blob/f6a75904cbf4979be6213034e1459ff1c397a948/packages/block-editor/src/components/copy-handler/index.js#L161-L162)
1. Select a range of rich text within a single block and copy with <kbd>Meta+C</kbd> — handled by [useCopyHandler](https://github.com/WordPress/gutenberg/blob/f6a75904cbf4979be6213034e1459ff1c397a948/packages/rich-text/src/component/use-copy-handler.js#L39-L45)
1. Click the Copy All Content button in the More menu — handled by [CopyContentMenuItem](https://github.com/WordPress/gutenberg/blob/f6a75904cbf4979be6213034e1459ff1c397a948/packages/edit-post/src/plugins/copy-content-menu-item/index.js#L26)
1. Select a block and click the Copy button in the block's dropdown menu — handled by [CopyMenuItem](https://github.com/WordPress/gutenberg/blob/f6a75904cbf4979be6213034e1459ff1c397a948/packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js#L43)
1. (anything else that I'm forgetting?)

The first two cases are those in which multiple types of data are added to the clipboard (`text/plain`, `text/html`, and some special rich-text-related types), while in cases 3 and 4 we only set `text/plain` with serialised block markup.

Case 1 is the one that this PR is changing. Note that, in case 2, the plain-text payload was already free of the HTML-comment block demarcations, so this PR can be said to make copying more consistent across the editor.

@draganescu: Can we list pros and cons of stripping content down to the visible plain text?

mcsf commented:
Years ago, in a discussion in this repo around copying and pasting, someone suggested some sort of just-in-time UI, like a popover around the cursor, asking the user what they would like to copy or paste: the plain text, the rich text, or the markup. 

I don't really know how I would find that thread nowadays; I don't think that anything came of it, but I think there's an opportunity for better clipboard ergonomics.

In my mind, it's similar to this URL-opening tool called Choosy, but reacting to <kbd>Meta+C</kbd> or <kbd>Meta+V</kbd>:


https://user-images.githubusercontent.com/150562/170766291-3fcd2082-aa05-44a6-a9fe-f22bcead016b.mp4

mtias commented:
> Years ago, in a discussion in this repo around copying and pasting, someone suggested some sort of just-in-time UI, like a popover around the cursor

Me! Ulysses does something similar.

draganescu commented:
Wow, we're handling the clipboard in so many places 😁 !

> Can we list pros and cons of stripping content down to the visible plain text?

#### Pros

* **Pro**: the user pastes visible text in apps that expect to receive just that, visible text. It's hard to argue that I copy visible text in WordPress paste in Notepad and _expect_ a soup of JSON and HTML, plus the visible text.
* **Pro**: we are not lying to the computer, if we say we set `text/plain`  then this is what we place there.
* **Pro**: we're not encouraging manual editing on block HTML. When we paste in an app that takes rich text or HTML, the user visually edits the content and when it pastes back to Gutenberg, we can trust our platform for the correct transformation. With block markup **and** HTML in plain text we have no guarantees over what we're getting, we don't have a super solid auto fixing system and even what is wrong is hard to highlight. So better safe than sorry.
* **Pro**: it matches other systems and tools - pasting rich media and rich formatting text from Google Docs or MS Word in a plain text app results in visible text being pasted, nothing else.

#### Cons

* **Con**: may be confusing if unexpected - pasting plain text not only loses the formatting but also the media. This would only happen if the capabilities of the destination application are unknown to the user, which is not for WP to solve.
* **Con**: some applications don't handle clipboard objects well and default to `text/plain` _if that is preset_. This is a bug to be fixed in those applications.

>  a popover around the cursor, asking the user what they would like to copy or paste: the plain text, the rich text, or the markup.

MS Word has the Paste Special command which opens an ancient popup window allowing the user to select what to paste (HTML, formatted, unformatted). 

>  identifying use cases

To make things consistent I think we should:

(a) make the behavior of:
* Select a range of blocks and copy with Meta+C
* Select a range of rich text within a single block and copy with Meta+C
* Select a block and click the Copy button in the block's dropdown menu

... be consistent in putting visible text in `text/plain` and html + block markup in `text/html`, plus the formatted rich text.

(b) rename the more menu's "Copy All Content" to "Copy Block Content" and override `text/plain` with block markup and HTML in this special situation, which would be a shortcut to switching to code view and copying all visible text from there.

mcsf commented:
I think your arguments and proposals are fair, @draganescu. I'd like an extra opinion, perhaps from @ellatrix?

ellatrix commented:
This is fine with me. Ideally, the plain text type should be plain text, the HTML type should be just HTML without block demarkations, and we should use a custom type for the proper serialised output with the block demarkations.

I'm pretty sure there's at least one existing issue that can be closed by this PR?

draganescu commented:
@mcsf @ellatrix so what's next for this? Can it be merged like this, or should I build the following:

(a) make the behavior of:

    Select a range of blocks and copy with Meta+C
    Select a range of rich text within a single block and copy with Meta+C
    Select a block and click the Copy button in the block's dropdown menu

... be consistent in putting visible text in text/plain and html + block markup in text/html, plus the formatted rich text.

(b) rename the more menu's "Copy All Content" to "Copy Block Content" and override text/plain with block markup and HTML in this special situation, which would be a shortcut to switching to code view and copying all visible text from there.

... on top of this?

ellatrix commented:
> plus the formatted rich text

Not sure what you mean here.
Yeah, everything should be consistent in all our copy events
Should we add e2e tests?

mcsf commented:
If I understood Ella correctly, the part about allowing users to copy HTML _stripped of block demarcations_ into `text/html` is something to consider after this PR, not in it. But, as for the rest, I believe it should all be in this PR:

- copy consistency per [item (a)](https://github.com/WordPress/gutenberg/pull/41366#issuecomment-1140935092)
- e2e tests
- (optionally here, or subsequently) renaming of the Copy All Content button per [item (b)](https://github.com/WordPress/gutenberg/pull/41366#issuecomment-1140935092)

draganescu commented:
Some complexity here :)

<img width="980" alt="Screenshot 2022-06-10 at 19 14 20" src="https://user-images.githubusercontent.com/107534/173108162-8cff8aae-ad25-461a-90e0-e271717035ba.png">

draganescu committed:
rename the copy block data menu item to hopefully stress that the block markup is copied as well

draganescu committed:
edit block tool copy and more tools copy to refer to block

draganescu committed:
mend

draganescu committed:
mend

draganescu committed:
mend

draganescu commented:
@mcsf for consistency the following have been added:

- I added a small UI label improvement to block tools, to use plural "blocks" if multiple are selected
- I changed the editor (post and site) label for "Copy content" to "Copy all blocks" which is to make a connection that the block tools copy is similar to editor tools copy, and results in block markup being copied as plain text.
- I updated the plain text setter for `useClipboardHandler` to remove multiple consecutive new lines. I opted to have it just here and not update the exposed `stripHtml` api.
- I have added a test to guard against pasting block content in plain text contexts, by using a code block. I think it work for our purpose.

mcsf commented on packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js:
```diff
@@ -41,7 +41,9 @@ const POPOVER_PROPS = {
 function CopyMenuItem( { blocks, onCopy } ) {
 	const ref = useCopyToClipboard( () => serialize( blocks ), onCopy );
-	return <MenuItem ref={ ref }>{ __( 'Copy' ) }</MenuItem>;
+	const copyMenuItemLabel =
+		blocks.length > 1 ? __( 'Copy blocks' ) : __( 'Copy block' );
```
Should never attempt to manually wrangle plurals in i18n. Instead, use `_n`:

```suggestion
	const copyMenuItemLabel = _n( 'Copy block', 'Copy blocks', blocks.length );
```

Different languages handle plurals and counting in ways that exceed our imagination, like:
- singular (1), dual (2), plural (3+)
- no inflection at all
- special zero case
- counting particles
- etc.

draganescu replied:
I had no idea abou `_n`, now I do!

mcsf committed:
Handle BR tags and trim whitespace when producing plain text

mcsf commented on packages/block-editor/src/components/copy-handler/index.js:
```diff
@@ -213,6 +217,23 @@ function CopyHandler( { children } ) {
 	return <div ref={ useClipboardHandler() }>{ children }</div>;
 }
 
+/**
+ * Given a string of HTML representing serialized blocks, returns the plain
+ * text extracted after stripping the HTML of any tags and fixing line breaks.
+ *
+ * @param {string} html Serialized blocks.
+ * @return {string} The plain-text content with any html removed.
+ */
+function toPlainText( html ) {
+	// Manually handle BR tags as line breaks prior to `stripHTML` call
+	html = html.replace( /<br>/g, '\n' );
+
+	const plainText = stripHTML( html ).trim();
+
+	// Merge any consecutive line breaks
+	return plainText.replace( /\n\n+/g, '\n\n' );
+}
+
```
I took the liberty of pushing this change to improve the produced plain text.

In my testing I was copying this partial content across two paragraphs:

> abc
>
> def **ghi** jkl
> _abcdefghi_

**Before**, the resulting plain text was:

```
abc
def ghi jklabcdefghi
```

**Now**:

```
abc

def ghi jkl
abcdefghi
```

mcsf replied:
I grepped around the `raw-handling` menu to see if I could/should reuse anything to clean up the HTML while preserving line breaks, but didn't find anything that felt appropriate, hence this little dedicated function. (cc @ellatrix in case you have a better idea)

draganescu replied:
Super awesome 👏🏻

mcsf reviewed (approved):
I think this just needs the `_n` fix (feel free to commit the suggestion from GH) and is good to go.

draganescu committed:
use `_n` for plural of block

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>

draganescu commented:
:D `eferenceError: _n is not defined` I felt too free to commit from GH.

mcsf commented:
> :D `eferenceError: _n is not defined` I felt too free to commit from GH.

What, it doesn't run? Pff, minor detail! 😅

draganescu committed:
import _n

draganescu committed:
update snapshot with lesst newlines

gaambo commented:
FYI: I worried that this change might break the possibility to copy multiple blocks / nested blocks to save them in a .html file for including as a pattern in a theme. But using the "Copy Block" menu button and pasting it into VSCode or Windows Notepad also pastes the block markup - so everything's fine 👍
